monitoring.py

```diff
@@ -233,48 +232,60 @@ def _reset(self):
                                         0, timestamp=ts)
         
     def _update(self):
+        self.log.debug(f"StorageLogger: Updating storage usage in {self.directory}.")
         try:
             current_files = glob.glob(os.path.join(self.directory, '*'))
             current_files.sort()    # The files should have sensible names that
                                     # reflect their creation times
             
+            new_files, new_file_sizes = deque(), deque()
             for filename in current_files:
-                if filename not in self._files:
-                    filesize = getsize(filename)
-                    self._files.append(filename)
-                    self._file_sizes.append(filesize)
+                try:
+                    i = self._files.index(filename)
+                    new_files.append(filename)
+                    new_file_sizes.append(self._file_sizes[i])
+                except ValueError:
+                    size = getsize(filename)
+                    new_files.append(filename)
+                    new_file_sizes.append(size)
         except Exception as e:
             self.log.warning("Quota manager could not refresh the file list: %s", str(e))
-            
+        self._files = new_files
+        self._file_sizes = new_file_sizes
+ 
     def _halt(self):
         self._reset()
         
     def _manage_quota(self):
         t0 = time.time()
         total_size = sum(self._file_sizes)
         
-        i = 0
-        removed = 0
-        removed_size = 0
+        to_remove = []
+        to_remove_size = 0
         while total_size > self.quota and len(self._files) > 1:
-            try:
-                subprocess.check_call(['rm', '-rf', self._files[i]],
-                                       stdout=subprocess.DEVNULL,
-                                       stderr=subprocess.DEVNULL)
-                removed += 1
-                removed_size += self._file_sizes[i]
-                total_size -= self._file_sizes[i]
-                
-                del self._files[i]
-                del self._file_sizes[i]
-            except Exception as e:
-                self.log.warning("Quota manager could not remove '%s': %s", to_remove, str(e))
-                i += 1
-                
-        if removed > 0:
+            fn = self._files.pop()
+            if (len(fn) <= len(self.directory)) or \
+                (not fn.startswith(self.directory)) or \
+                    (len(fn) <= MINIMUM_TO_DELETE_PATH_LENGTH):
```

Comment:
I don't follow this series of checks on the filename.

Reply:
This is me being real paranoid. This makes sure that the fn (i.e. path) is under `self.directory` and in case `self.directory` was set incorrectly, it will be less possible for other data files to get deleted (say someone put `/` in `self.directory` or something).